### ISSUE TITLE
QB64 should offer built-in support for HTTP and HTTPS

### ISSUE THREAD

**Issue by mkilgore** [QB64-Phoenix-Edition/QB64pe#46]
The current QB64 networking only extends to basic TCP streams and UDP. This is certainly useful, but it's not all that usable for a lot of basic tasks people actually want to do with networking, in particular HTTP and HTTPS. Writing your own HTTP handler in QB64 is fairly hard to actually accomplish, and frankly not something you should really have to do. HTTPS is also effectively impossible since it would require implementing TLS support yourself. The best option is simply pulling in external libraries via `DECLARE LIBRARY`. The QB64 IDE itself just shells out to `curl` to download Wiki pages, effectively giving up on the idea of doing it itself.

The basics of doing this is really not that hard, as `libcurl` is MIT licensed and can do all of this with a fairly simple interface. The hard part would be the interface offered, and whether it can be added into the existing networking functions (`_OPENCLIENT()` and friends) or requires some new ones.

**Comment by mkilgore** [comment]
This has come up a lot. and after looking into `libcurl` I think it's decently straight-forward to get a basic version of this working. Going along with my suggestion in #21 I'll hide the initial version behind an `$UNSTABLE:HTTP` type of command while we determine what the interface should look like. I think we can make it work with the existing `_OPENCLIENT()` interface with some extensions for HTTP-specific settings, but we'll have to see. I don't think those extensions are necessary for the initial version, but if it's too hard to integrate the HTTP settings (cookies, headers, etc.) then a new API may be warranted.

**Comment by mkilgore** [comment]
This is just for downloading. not rendering or anything along those lines. It's intended to replace the various existing solutions such as shelling out to `curl` or writing your own HTTP logic. Both cases have pretty large drawbacks (shelling to `curl` has a lot of overhead and is not always available, custom written HTTP logic tends to be quite fragile and cannot support HTTPS), and using HTTP is such a common use case that you really shouldn't need to come up with such a custom solution, which is the idea behind this issue.

Unfortunately I haven't haven't gotten to work on this in a while, but the backend part for `libqb` is complete. Once it's done you'll be able to just do `_OPENCLIENT("https://example.com")` to talk to HTTP endpoints without needing any special logic. Eventually we'll likely need support for handling HTTP things like supply headers, cookies, etc. but I don't think that's necessary for the initial version.

**Comment by Spriggsy95** [comment]
Well I'm quite familiar with the Windows side of HTTPS. If we could use the Win32 functions for this rather than libcurl on Windows then that would be sweet. But if you know libcurl well enough then we can just go with that.

**Comment by SpriggsySpriggs** [comment]
I'm quite familiar with HTTPS in Windows using Win32 API so I might be able to help out there. However, if you'd rather we go with libcurl then we can

**Comment by mkilgore** [comment]
I think it's worth it to just use `libcurl` on all platforms to keep the code the same, and also I already wrote that part of the code :D I just need to finish up the changes to the QB64 networking commands so that they call into the new HTTP logic.

**Comment by SpriggsySpriggs** [comment]
Nice! I look forward to trying out the https functionality when it is available for testing.

**Comment by mkilgore** [comment]
Still working on this, thought I'd provide a bit of an update (and am also interested in getting some feedback. I more or less have `_OPENCLIENT` and friends working fine with HTTP, the only issue is that the API itself is fairly limited compared to what you should be able to do with HTTP.

So with that said, like I think I mentioned earlier, I'm looking at what 'proper' HTTP support would look like, this is what I'm currently thinking:

```vb
httpHandle& = _HttpOpen([string url]) ' This creates a HTTP request. It does not start it. The URL can optionally be provided here

_HttpSetRequestUrl httpHandle&, string url

_HttpSetRequestVerb httpHandle&, string verb ' GET is the default, can be changed to Ex. `POST`, `PUT`, etc.

_HttpAddRequestHeader httpHandle&, string name, string contents ' Add a header of 'name' with 'contents'

_HttpSetRequestBody httpHandle&, string body ' Adds body contents for the request, Ex. POST data

statusCode& = _HttpSendRequest(httpHandle&) ' This starts the request and returns the status code. -1 is returned on error

headers& = _HttpResponseHeaderCount(httpHandle&) ' total number of headers on response
headerName$ = _HttpResponseHeaderName(httpHandle&, 1) ' Get the first header's name
headerContents$ = _HttpResponseHeaderContents(httpHandle&, 1) ' Get the first header's contents

length& = _HttpContentLength(httpHandle&) ' Returns the content-length. -1 if no length was given.
closed& = EOF(httpHandle&) ' Returns whether the HTTP response is finished (meaning all data has been read)

' Reads the current string contents we have. May not be the entire message of the server is still sending data
GET #httpHandle&, , content$

' Typical loop to read entire response in one go
contents$ = ""
DO WHILE NOT EOF(httpHandle&)
    _LIMIT 100
    GET #httpHandle&, , s$
    IF LEN(s$) THEN contents$ = contents$ + s$
LOOP

' Release the HTTP request we opened
CLOSE #httpHandle&
```

I think it's a little too wordy at the moment, that can probably be worked on. I think there's some potential for a set of commands more like `_Http #httpHandle& Set Verb "verb"` instead of the typical sub/function.

The API above is fairly low-level but it's intended to let you control everything about the HTTP request. That said it's easy to wrap it and provide versions (either as extra commands, or as provided QB64 code) that are simpler for the typical use-case. Ex:

```vb
FUNCTION Download$(string url, long statusCode)
    httpHandle& = _HttpOpen(url)
    statusCode = _HttpSendRequest(httpHandle&)

    IF statusCode <> 200 THEN EXIT FUNCTION

    content$ = ""
    DO WHILE NOT EOF(httpHandle&)
        _LIMIT 100
        GET #httpHandle&, , s$
        IF LEN(s$) THEN content$ = content$ + s$
    LOOP

    Download$ = content$
END FUNCTION
```

**Comment by SpriggsySpriggs** [comment]
This looks amazing. This is basically what I do with my Win32 stuff but I have to first make so many declarations and wrappers and whatnot. If the final product is anything like this sample you've shared then I'm sold.

**Comment by RhoSigma-QB64** [comment]
Can't hardly wait to see this finished and released, but:
"Be patient and give things the time they need to turn into something good."
That's what my grandma always said.
